rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/event/RMInitialState.java

```diff
@@ -55,82 +61,145 @@
-                break;
-            }
-        }
+    public long getLatestCounter() {
+        return latestCounter.get();
     }
 
-    public void nodeRemoved(RMNodeEvent removedEvent) {
-        Iterator<RMNodeEvent> events = nodesList.iterator();
-        while (events.hasNext()) {
-            RMNodeEvent nodeEvent = events.next();
-            if (removedEvent.getNodeUrl().equals(nodeEvent.getNodeUrl())) {
-                events.remove();
-                break;
-            }
-        }
+    public void nodeAdded(RMNodeEvent event) {
+        updateCounter(event);
+        nodeEvents.put(event.getNodeUrl(), event);
+
     }
 
-    public void nodeAdded(RMNodeEvent event) {
-        nodesList.add(event);
+    public void nodeStateChanged(RMNodeEvent event) {
+        updateCounter(event);
+        nodeEvents.put(event.getNodeUrl(), event);
+    }
+
+    public void nodeRemoved(RMNodeEvent event) {
+        updateCounter(event);
+        nodeEvents.put(event.getNodeUrl(), event);
     }
 
-    public void nodeSourceStateChanged(RMNodeSourceEvent stateChangedEvent) {
-        boolean existNodeSource = false;
-        int size = nodeSources.size();
-        for (int i = 0; i < size; i++) {
-            if (stateChangedEvent.getSourceName().equals(nodeSources.get(i).getSourceName())) {
-                existNodeSource = true;
-                nodeSources.set(i, stateChangedEvent);
-                break;
+    public void nodeSourceAdded(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.put(event.getSourceName(), event);
+    }
+
+    public void nodeSourceRemoved(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.put(event.getSourceName(), event);
+    }
+
+    public void nodeSourceStateChanged(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.put(event.getSourceName(), event);
+    }
+
+    private void updateCounter(RMEvent event) {
+        latestCounter.set(Math.max(latestCounter.get(), event.getCounter()));
+    }
+
+    /**
+     * Clones current state events, but keep only those events which has counter bigger than provided 'filter'
+     * Event counter can take values [0, +).
+     * So if filter is '-1' then all events will returned.
+     * @param filter
+     * @return rmInitialState where all the events bigger than 'filter'
+     */
+    public RMInitialState cloneAndFilter(long filter) {
+        long actualFilter;
+        if (filter <= latestCounter.get()) {
+            actualFilter = filter;
+        } else {
+            LOGGER.info(String.format("Client is aware of %d but server knows only about %d counter. " +
+                                      "Probably because there was network server restart.",
+                                      filter,
+                                      latestCounter.get()));
+            actualFilter = -1; // reset filter to default  value
```

Comment:
We can create a constant for -1.
Maybe we can also use this constant in `@DefaultValue("-1")`? I doubt it.